Fundamental test coverage of userdb module reusing existing structure.

jonasbardino opened the pull request:
Fundamental test coverage of `userdb` module reusing existing structure.

Some tests are disabled on purpose because changes are needed in the tested module. Marked as follow-up.

jonasbardino committed:
Fundamental test coverage of userdb module reusing existing structure.

jonasbardino committed:
Increase coverage but with a number of tests requiring tweaks to the underlying
`userdb` module and in particular the locking from `fileio`.

albu-diku commented on tests/test_mig_shared_userdb.py:
```diff
@@ -0,0 +1,371 @@
+    def test_lock_unlock_user_db(self):
+        """Test lock/unlock cycle for user database"""
+        # Exclusive locking
+        flock = lock_user_db(self.user_db_path, exclusive=True)
```
I think there is a separate establishes an exclusive lock test here.

jonasbardino replied:
Split the joint tests of exclusive and non-exclusive lock+unlock into individual tests.

albu-diku commented on tests/test_mig_shared_userdb.py:
```diff
@@ -0,0 +1,371 @@
+            save_user_db("invalid content", self.user_db_path)
+
+    def test_load_user_db_thread_safety(self):
+        """Test thread-safe loading with shared lock"""
```
Could you clarify how this test is chcking thread safety? I'd have expected to see threads being spawned for this.

jonasbardino replied:
It emulates another thread concurrently locking the database non-exclusively and verifies that it does not prevent the main one from loading the user database.

albu-diku reviewed (changes requested):
Found some things in a first pass that I believe need to be addressed.

jonasbardino committed:
Split joint test for non-exclusive and exclusive lock+unlock into two
individual tests as suggested in PR review.

jonasbardino commented on tests/test_mig_shared_userdb.py:
```diff
@@ -0,0 +1,371 @@
+        unlock_user_db(flock)
+
+        # Shared locking
+        flock = lock_user_db(self.user_db_path, exclusive=False)
```
Split the joint tests of exclusive and non-exclusive lock+unlock into individual tests.

jonasbardino committed:
Address review feedback:
 * remove excess user db load in save test
 * rename pickle test to explain abi check and split into load and save test
 * fork a `_generate_sample_db` helper from `_create_sample_db` for abi test
 * clarify how thread-safety test emulates concurrent multi-threaded access

jonasbardino commented on tests/test_mig_shared_userdb.py:
```diff
@@ -0,0 +1,371 @@
+        """Test saving user database content"""
+        sample_db = self._create_sample_db()
+        try:
+            loaded = load_user_db(self.user_db_path)
```
Right, removed.

jonasbardino commented on tests/test_mig_shared_userdb.py:
```diff
@@ -0,0 +1,371 @@
+        self.assertFalse(delayed_thread.is_alive())
+
+    def test_pickle_roundtrip(self):
+        """Verify pickle serialization compatibility"""
```
It's perhaps a bit of a stretch but in case we should ever naively decide to change `save` and `load` db functions to e.g. use another more efficient pickle protocol this test will scream at us for breaking backwards compatibility - unlike at test with save+load.
I've renamed the test and introduced a matching `pickle.dump` + `load` test.

jonasbardino committed:
Address more review feedback and switch more tests away from checking results
indirectly with API calls to direct checking e.g. using pickle load and dump.
Split joint testing of several load scenarios into individual tests to ease
regression tracking. Add a save+load cycle integration test instead. Add save
empty db test.

jonasbardino commented on tests/test_mig_shared_userdb.py:
```diff
@@ -0,0 +1,371 @@
+        empty_db = {}
+        save_user_db(empty_db, self.user_db_path)
+        try:
+            loaded = load_user_db(self.user_db_path)
```
Agreed and split multi-test into individual tests using direct pickle calls for saving.

jonasbardino commented:
> Found some things in a first pass that I believe need to be addressed.

Thanks for reviewing. I've made another pass to address your feedback so I'll request another review.

albu-diku commented on tests/test_mig_shared_userdb.py:
```diff
@@ -0,0 +1,396 @@
+        with self.assertRaises(TypeError):
+            save_user_db("invalid content", self.user_db_path)
+
+    def test_load_user_db_thread_safety(self):
```
I still find the name here rather uncomfortable. Could we consider renaming this "test_load_uesr_db_allow_conurrent_access()" or something along those lines and drop the mention of threads? Seeing that word continues to be jarring when there are no uses of actual threads in the test itself.

jonasbardino replied:
Yes, adjusted name as suggested.

albu-diku reviewed (approved):
I've left a comment about the naming of one of these tests, I tihnk it would be a small change that would benefit the reader. Bu it's not enough to block what's here, hence adding my approval.

jonasbardino committed:
Adjust test case name as suggested in review (thanks @albu-diku).
